src/operator/tensor/elemwise_binary_broadcast_op.h

```diff
@@ -196,47 +166,38 @@ void ReduceToAssign(mshadow::Tensor<xpu, 2, DType> out, const OpReqType req, con
-                                               const std::vector<TBlob>& outputs,
-                                               const TShape& new_lshape,
-                                               const TShape& new_rshape,
-                                               const TShape& new_oshape) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  Tensor<xpu, ndim, DType> ograd =
-    inputs[0].get_with_shape<xpu, ndim, DType>(new_oshape.get<ndim>(), s);
-  Tensor<xpu, ndim, DType> lgrad =
-    outputs[0].get_with_shape<xpu, ndim, DType>(new_lshape.get<ndim>(), s);
-  Tensor<xpu, ndim, DType> rgrad =
-    outputs[1].get_with_shape<xpu, ndim, DType>(new_rshape.get<ndim>(), s);
-  ReduceToAssign<red::sum>(lgrad, req[0], F<LOP>(ograd));
-  ReduceToAssign<red::sum>(rgrad, req[1], F<ROP>(ograd));
-}
-
 template<typename xpu, typename LOP, typename ROP>
 void BinaryBroadcastBackwardUseNone(const nnvm::NodeAttrs& attrs,
                                     const OpContext& ctx,
                                     const std::vector<TBlob>& inputs,
                                     const std::vector<OpReqType>& req,
                                     const std::vector<TBlob>& outputs) {
+  // LOG(INFO) << attrs.name;
+  using namespace broadcast;
   TShape new_lshape, new_rshape, new_oshape;
-  bool need_bc = BinaryBroadcastShapeCompact(outputs[0].shape_, outputs[1].shape_, inputs[0].shape_,
+  int ndim = BinaryBroadcastShapeCompact(outputs[0].shape_, outputs[1].shape_, inputs[0].shape_,
                                              &new_lshape, &new_rshape, &new_oshape);
-  if (!need_bc) {
+  if (!ndim) {
     BinaryBackwardUseNone<xpu, LOP, ROP>(attrs, ctx, inputs, req, outputs);
   } else {
     MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-      if (new_oshape.ndim() == 2) {
-        BinaryBroadcastBackwardUseNoneImpl<xpu, 2, DType, LOP, ROP>(
-          ctx, inputs, req, outputs, new_lshape, new_rshape, new_oshape);
-      } else {
-        BinaryBroadcastBackwardUseNoneImpl<xpu, MXNET_SPECIAL_MAX_NDIM, DType, LOP, ROP>(
-          ctx, inputs, req, outputs, new_lshape, new_rshape, new_oshape);
-      }
+      Stream<xpu> *s = ctx.get_stream<xpu>();
+      CTensor<DType> ograd =
+        inputs[0].get_with_shape<xpu, MAX_DIM, DType>(new_oshape.get<MAX_DIM>(), s);
+      CTensor<DType> lgrad =
+        outputs[0].get_with_shape<xpu, MAX_DIM, DType>(new_lshape.get<MAX_DIM>(), s);
+      CTensor<DType> rgrad =
+        outputs[1].get_with_shape<xpu, MAX_DIM, DType>(new_rshape.get<MAX_DIM>(), s);
+      // Request temporary storage
+      size_t workspace_size_l = ReduceWorkspaceSize<red::sum, DType, LOP>(s, lgrad, req[0], ograd);
+      size_t workspace_size_r = ReduceWorkspaceSize<red::sum, DType, ROP>(s, rgrad, req[1], ograd);
+      size_t workspace_size = workspace_size_l + workspace_size_r;
+      Tensor<xpu, 1, char> workspace =
```

Comment:
I didn't go with atomic methods since there could be combinations of operations/data types for which no hardware atomic method exists